### ISSUE TITLE
Dimensionality problems in SquareLossLayer?

### ISSUE THREAD

**Issue by philtomson** [pluskid/Mocha.jl#51]
We've had some suspicions about the SquareLossLayer perhaps not giving us correct answers so I went into the SquareLossLayer code and replace the BLAS operations with:
    state.loss = (0.5/get_num(pred))*(vecnorm(label.data - state.pred_copy.data))^2
just to see if that would give different answers.

The result was a dimensionality mismatch error at that line.

I figured out how to use Debug and set a breakpoint at the line prior and checked the dimensions:

debug:66> size(label.data)
(750,750,10)  (our image data size is 750x750 and the 10 is the batch size)
debug:66> size(state.pred_copy.data)
(750,750,1,10) (750x750 image, 1 example, batchsize 10)

The BLAS functions AFAIKT don't do any kind of dimension check so I'm wondering if we're actually getting an incorrect answer as the underlying BLAS C code just moves the pointer through memory?

**Comment by pluskid** [comment]
Hi Phil,


Tensor of size (750,750,1,10) and (750,750,10) should have the same underlying memory layout. I believe the blas should not have problem here. But if you want to check, you could reshape one tensor to the same shape as the other before doing subtraction. Or  alternatively just convert both to a vector via the vec function.






- chiyuan



​Send from my phone



On Wednesday, Feb 4, 2015 at 4:24 PM, Phil Tomson <notifications@github.com>, wrote:

We've had some suspicions about the SquareLossLayer perhaps not giving us correct answers so I went into the SquareLossLayer code and replace the BLAS operations with:

    state.loss = (0.5/get_num(pred))*(vecnorm(label.data - state.pred_copy.data))^2

just to see if that would give different answers.


The result was a dimensionality mismatch error at that line.


I figured out how to use Debug and set a breakpoint at the line prior and checked the dimensions:


debug:66> size(label.data)

(750,750,10)  (our image data size is 750x750 and the 10 is the batch size)

debug:66> size(state.pred_copy.data)

(750,750,1,10) (750x750 image, 1 example, batchsize 10)


The BLAS functions AFAIKT don't do any kind of dimension check so I'm wondering if we're actually getting an incorrect answer as the underlying BLAS C code just moves the pointer through memory?

**Comment by philtomson** [comment]
Yes, can confirm very similar answers if I use the pure julia (with vec) and the BLAS. Can close issue, thanks.
